### ISSUE TITLE
chapter 9 - regularization.py code

### ISSUE THREAD

**Issue by chopsuey5000** [PyImageSearch/Deep-Learning-for-Computer-Vision-with-Python#2]
* OS Platform: windows 10
* Python version: 3.7.10
* OpenCV version: 4.5.1.48
* TensorFlow version: None
* Scikit-learn: 0.24.1

## Issue

I'm running the regularization.py code with 10 iterations and I get a warning saying (ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.):

```
python regularization.py --dataset ../datasets/animals
[INFO] loading images...
[INFO] processed 500/3000
[INFO] processed 1000/3000
[INFO] processed 1500/3000
[INFO] processed 2000/3000
[INFO] processed 2500/3000
[INFO] processed 3000/3000
[INFO] training model with 'None' penalty
C:\Users\chops\anaconda3\envs\DL4CV_SB\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:577: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO] 'None' penalty accuracy: 49.33%
[INFO] training  model with 'l1' penalty
C:\Users\chops\anaconda3\envs\DL4CV_SB\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:577: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO] 'l1' penalty accuracy: 45.47%
[INFO] training model with 'l2' penalty
C:\Users\chops\anaconda3\envs\DL4CV_SB\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:577: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO] 'l2' penalty accuracy: 50.53%
```

Then I increased the number of iterations until 70 and the warning have disappeared but I cannot see any accuracy improvement using L1 or L2 regularizations as the textbook mentions.

```
>python regularization.py --dataset ../datasets/animals
[INFO] loading images...
[INFO] processed 500/3000
[INFO] processed 1000/3000
[INFO] processed 1500/3000
[INFO] processed 2000/3000
[INFO] processed 2500/3000
[INFO] processed 3000/3000
[INFO] training model with 'None' penalty
[INFO] 'None' penalty accuracy: 51.33%
[INFO] training model with 'l1' penalty
[INFO] 'l1' penalty accuracy: 47.47%
[INFO] training model with 'l2' penalty
[INFO] 'l2' penalty accuracy: 48.53%
```

Could please tell what is the issue?

Thanks in advance,

**Comment by sayakpaul** [comment]
Can you one of our VMs and see if this issue persists?

**Comment by chopsuey5000** [comment]
Hi Sayak,

I've tried with the VM and the problem persists.

With 10 iterations the output is:

```
$ python regularization.py --dataset ../datasets/animals
[INFO] loading images...
[INFO] processed 500/3000
[INFO] processed 1000/3000
[INFO] processed 1500/3000
[INFO] processed 2000/3000
[INFO] processed 2500/3000
[INFO] processed 3000/3000
[INFO] tranining model with 'None' penalty
[INFO] 'None' penalty accuracy: 53.73%
[INFO] tranining model with 'l1' penalty
[INFO] 'l1' penalty accuracy: 53.33%
[INFO] tranining model with 'l2' penalty
[INFO] 'l2' penalty accuracy: 54.53%
(dl4cv) pyimagesearch@pyimagesearch-dl4cv:~/Documents/deep_learning/c9_optimization_methods$ python regularization.py --dataset ../datasets/animals
[INFO] loading images...
[INFO] processed 500/3000
[INFO] processed 1000/3000
[INFO] processed 1500/3000
[INFO] processed 2000/3000
[INFO] processed 2500/3000
[INFO] processed 3000/3000
[INFO] tranining model with 'None' penalty
/home/pyimagesearch/.virtualenvs/dl4cv/lib/python3.6/site-packages/sklearn/linear_model/stochastic_gradient.py:561: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO] 'None' penalty accuracy: 45.07%
[INFO] tranining model with 'l1' penalty
/home/pyimagesearch/.virtualenvs/dl4cv/lib/python3.6/site-packages/sklearn/linear_model/stochastic_gradient.py:561: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO] 'l1' penalty accuracy: 43.07%
[INFO] tranining model with 'l2' penalty
/home/pyimagesearch/.virtualenvs/dl4cv/lib/python3.6/site-packages/sklearn/linear_model/stochastic_gradient.py:561: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)
[INFO] 'l2' penalty accuracy: 51.87%
```

And with 100 iterations the output is:

```
$ python regularization.py --dataset ../datasets/animals
[INFO] loading images...
[INFO] processed 500/3000
[INFO] processed 1000/3000
[INFO] processed 1500/3000
[INFO] processed 2000/3000
[INFO] processed 2500/3000
[INFO] processed 3000/3000
[INFO] tranining model with 'None' penalty
[INFO] 'None' penalty accuracy: 53.73%
[INFO] tranining model with 'l1' penalty
[INFO] 'l1' penalty accuracy: 53.33%
[INFO] tranining model with 'l2' penalty
[INFO] 'l2' penalty accuracy: 54.53%
```

**Comment by sayakpaul** [comment]
I see. What happens when you up it 1000?

**Comment by chopsuey5000** [comment]
Same result as with 100

```
$ python regularization.py --dataset ../datasets/animals
[INFO] loading images...
[INFO] processed 500/3000
[INFO] processed 1000/3000
[INFO] processed 1500/3000
[INFO] processed 2000/3000
[INFO] processed 2500/3000
[INFO] processed 3000/3000
[INFO] tranining model with 'None' penalty
[INFO] 'None' penalty accuracy: 53.73%
[INFO] tranining model with 'l1' penalty
[INFO] 'l1' penalty accuracy: 53.33%
[INFO] tranining model with 'l2' penalty
[INFO] 'l2' penalty accuracy: 54.53%
```

**Comment by sayakpaul** [comment]
You meant 1000, right?

**Comment by chopsuey5000** [comment]
Yes

**Comment by sayakpaul** [comment]
I took a look at the original results again: 

![image](https://user-images.githubusercontent.com/22957388/115862471-d3f09e00-a451-11eb-9f98-d0478e089efc.png)

They don't vary that much from what you have gotten. The increased performance might just have been due to the stochasticity that is induced in the algorithm under consideration.
